### PR TITLE
feat: Add trait definition representation in DefCollector and HIR

### PR DESCRIPTION
**Pull request by yordanmadzhunkov**
# Description
This is the next PR from the series of Trait Implementation. I expected to be merged quickly, because it's some simple definition that will be used later

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by yordanmadzhunkov** [comment]
@jfecher

File: crates/noirc_frontend/src/hir/resolution/import.rs [hunk 1]
```diff
@@ -153,22 +153,22 @@ fn resolve_name_in_module(
...
         let found_ns = current_mod.find_name(segment);
+
         if found_ns.is_none() {
             return Err(PathResolutionError::Unresolved(segment.clone()));
         }
-
         // Check if it is a contract and we're calling from a non-contract context
         if current_mod.is_contract && !allow_contracts {
             return Err(PathResolutionError::ExternalContractUsed(segment.clone()));
         }
```
Comment (TomAFrench) [inline crates/noirc_frontend/src/hir/resolution/import.rs:171]:
```suggestion
        }

```

File: crates/noirc_frontend/src/hir/resolution/import.rs [hunk 2]
```diff
@@ -153,22 +153,22 @@ fn resolve_name_in_module(
...
         };
 
         current_mod = &def_maps[&new_module_id.krate].modules[new_module_id.local_id.0];
 
         // Check if namespace
         let found_ns = current_mod.find_name(segment);
+
         if found_ns.is_none() {
             return Err(PathResolutionError::Unresolved(segment.clone()));
         }
```
Comment (TomAFrench) [inline crates/noirc_frontend/src/hir/resolution/import.rs:167]:
```suggestion
        }

```

File: crates/noirc_frontend/src/hir/resolution/import.rs [hunk 3]
```diff
@@ -153,22 +153,22 @@ fn resolve_name_in_module(
...
             ModuleDefId::TypeAliasId(_) => panic!("type aliases cannot be used in type namespace"),
+            ModuleDefId::TraitId(id) => id.0,
             ModuleDefId::GlobalId(_) => panic!("globals cannot be in the type namespace"),
         };
 
         current_mod = &def_maps[&new_module_id.krate].modules[new_module_id.local_id.0];
 
         // Check if namespace
         let found_ns = current_mod.find_name(segment);
+
```
Comment (TomAFrench) [inline crates/noirc_frontend/src/hir/resolution/import.rs:164]:
```suggestion
        let found_ns = current_mod.find_name(segment);
```

**Review by TomAFrench — COMMENTED** [review]
~~Removing unrelated changes~~ Ah looks like I can't make changes to this PR as it's on a fork.

**Review by jfecher — CHANGES_REQUESTED** [review]
Looks good. We can merge once `Type::Trait` is removed.

File: crates/noirc_frontend/src/node_interner.rs [hunk 4]
```diff
@@ -61,6 +61,15 @@ pub struct NodeInterner {
...
     type_aliases: Vec<TypeAliasType>,
 
+    // Trait map.
+    //
+    // Each trait definition is possibly shared across multiple type nodes.
+    // It is also mutated through the RefCell during name resolution to append
+    // methods from impls to the type.
+    //
+    // TODO: We may be able to remove the Shared wrapper once traits are no longer types.
+    // We'd just lookup their methods as needed through the NodeInterner.
```
Comment (jfecher) [inline crates/noirc_frontend/src/node_interner.rs:71]: Noting that traits are not shared across multiple type nodes anymore and this TODO likely applies now

**Review by jfecher — APPROVED** [review]
I'm approving this to keep development speed high, but lets move the Trait struct to its own file (hir_def/traits?) and update the comment/type on the `traits` field of the NodeInterner as part of a later PR.

**Pull request merged**